smelter-core/src/pipeline/utils/input_buffer.rs

```diff
@@ -0,0 +1,198 @@
+
+#[derive(Clone)]
+pub(crate) enum InputBuffer {
+    // If input is required or has an offset do not add any buffer.
+    //
+    // - If input is required, buffering is not necessary.
+    // - If offset is in the future then extra buffering is not necessary
+    // - If offset in in the past, it already causing drops
+    None,
+    Const { buffer: Duration },
+    LatencyOptimized(Arc<Mutex<LatencyOptimizedBuffer>>),
+    Adaptive(Arc<Mutex<AdaptiveBuffer>>),
+}
+
+impl fmt::Debug for InputBuffer {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::None => write!(f, "None"),
+            Self::Const { buffer } => f.debug_struct("Const").field("buffer", buffer).finish(),
+            Self::LatencyOptimized(_) => f.debug_struct("LatencyOptimized").finish(),
+            Self::Adaptive(_) => f.debug_struct("Adaptive").finish(),
+        }
+    }
+}
+
+impl InputBuffer {
+    pub fn new(ctx: &PipelineCtx, opts: InputBufferOptions) -> Self {
+        match opts {
+            InputBufferOptions::None => InputBuffer::None,
+            InputBufferOptions::Const(buffer) => InputBuffer::Const {
+                buffer: buffer.unwrap_or(ctx.default_buffer_duration),
+            },
+            InputBufferOptions::LatencyOptimized => InputBuffer::LatencyOptimized(Arc::new(
+                Mutex::new(LatencyOptimizedBuffer::new(ctx)),
+            )),
+            InputBufferOptions::Adaptive => {
+                InputBuffer::Adaptive(Arc::new(Mutex::new(AdaptiveBuffer::new(ctx))))
+            }
+        }
+    }
+
+    pub fn pts_with_buffer(&self, pts: Duration) -> Duration {
+        match self {
+            InputBuffer::None => pts,
+            InputBuffer::Const { buffer } => pts + *buffer,
+            InputBuffer::LatencyOptimized(buffer) => buffer.lock().unwrap().pts_with_buffer(pts),
+            InputBuffer::Adaptive(buffer) => buffer.lock().unwrap().pts_with_buffer(pts),
+        }
+    }
+}
+
+/// Buffer intended for low latency inputs, if input stream is not delivers on time
+/// it quickly increasing. However when buffer is stable for some time it starts to shrink to
+/// minimize the latency.
+pub(crate) struct LatencyOptimizedBuffer {
+    sync_point: Instant,
+    /// We expect pts to be at least greater than sync_point.elapsed() + desired_buffer.
+    ///
+    /// This buffer should be large enough, so a packet can be decoded and
+    /// placed in queue before queue attempts to render that pts.
+    desired_buffer: Duration,
+
+    min_buffer: Duration,
+    max_buffer: Duration,
+
+    dynamic_buffer: Duration,
+
+    state: LatencyOptimizedBufferState,
+}
+
+impl LatencyOptimizedBuffer {
+    fn new(ctx: &PipelineCtx) -> Self {
+        Self {
+            sync_point: ctx.queue_sync_point,
+            desired_buffer: ctx.default_buffer_duration,
+            min_buffer: Duration::min(Duration::from_millis(20), ctx.default_buffer_duration),
```

Comment:
Those values might need to be adjusted, currentlly:
- target 80ms
- if above 130ms over 10 second then lower buffer
- if above 20ms but below 80ms increase buffer